Hyperbolic_surface_triangulation_2/include/CGAL/Hyperbolic_fundamental_domain_2.h

```diff
@@ -0,0 +1,210 @@
+The domain D is given as a convex geodesic hyperbolic polygon P given by the list of its vertices in the hyperbolic plane,
+together with a pairing of the sides of P, such that every two paired sides have the same length, and such that
+identifying every two paired sides in a way that respects the orientation of P would result in a closed
+orientable hyperbolic surface.
+*/
+template<class Traits>
+class Hyperbolic_fundamental_domain_2 {
+
+public:
+  typedef typename Traits::Hyperbolic_point_2                    Point;
+
+  Hyperbolic_fundamental_domain_2();
+  Hyperbolic_fundamental_domain_2(const std::vector<Point>& vertices, const std::vector<int>& pairings);
+
+  void set(const std::vector<Point>& vertices, const std::vector<int>& pairings);
+
+  int size() const; // Returns the number of vertices (equivalently, the number of sides)
+  Point vertex(int index) const; // Returns the index-th vertex
+  int paired_side(int index) const; // Returns the index of the side paired to side A, where A is the index-th side
+  Hyperbolic_isometry_2<Traits> side_pairing(int index) const;// Returns the isometry that maps side A to side B, where B is the index-th side, and A is the side paired to B
+
+  void from_stream(std::istream& s);
+  void to_stream(std::ostream& s) const;
+
+  bool is_valid() const;
+
+private:
+  std::vector<Point>                                               _vertices;
+  std::vector<int>                                                  _pairings;
+};
+
+template<class Traits> std::ostream& operator<<(std::ostream& s, const Hyperbolic_fundamental_domain_2<Traits>& domain);
+template<class Traits> void operator>>(std::istream& s, Hyperbolic_fundamental_domain_2<Traits>& domain);
+
+
+////////////////////////////////////////////////////////////////////////////////
+////////////////////////////////////////////////////////////////////////////////
+////////////////////////////////////////////////////////////////////////////////
+
+template<class Traits>
+Hyperbolic_fundamental_domain_2<Traits>::Hyperbolic_fundamental_domain_2() {}
+
+template<class Traits>
+Hyperbolic_fundamental_domain_2<Traits>::Hyperbolic_fundamental_domain_2(const std::vector<Point>& vertices, const std::vector<int>& pairings){
+  set(vertices, pairings);
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+template<class Traits>
+void Hyperbolic_fundamental_domain_2<Traits>::set(const std::vector<Point>& vertices, const std::vector<int>& pairings){
+  _vertices = vertices;
+  _pairings = pairings;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+template<class Traits>
+int Hyperbolic_fundamental_domain_2<Traits>::size() const{
+  return _vertices.size();
+}
+
+template<class Traits>
+typename Hyperbolic_fundamental_domain_2<Traits>::Point Hyperbolic_fundamental_domain_2<Traits>::vertex(int index) const{
```

Comment:
Use `const&` so that points do not get copied.

Reply:
done